app/Models/UserGroupEvent.php

```diff
@@ -38,39 +40,84 @@ public static function logGroupRename(?User $actor, Group $group, string $oldNam
     {
         return static::log($actor, static::GROUP_RENAME, $group, [
             'details' => [
-                'old_name' => $oldName,
+                'group_name' => null,
                 'new_name' => $newName,
+                'old_name' => $oldName,
             ],
         ]);
     }
 
-    public static function logUserAdd(?User $actor, User $user, Group $group): self
+    public static function logUserAdd(?User $actor, User $user, Group $group, array $modes): self
     {
         return static::log($actor, static::USER_ADD, $group, [
+            'details' => [
+                'modes' => $modes,
+                'user_name' => $user->username,
+            ],
+            'user_id' => $user->getKey(),
+        ]);
+    }
+
+    public static function logUserAddModes(?User $actor, User $user, Group $group, array $modes): self
+    {
+        return static::log($actor, static::USER_ADD_MODES, $group, [
+            'details' => [
+                'modes' => $modes,
+                'user_name' => $user->username,
+            ],
             'user_id' => $user->getKey(),
         ]);
     }
 
     public static function logUserRemove(?User $actor, User $user, Group $group): self
     {
         return static::log($actor, static::USER_REMOVE, $group, [
+            'details' => [
+                'user_name' => $user->username,
+            ],
             'user_id' => $user->getKey(),
         ]);
     }
 
+    public static function logUserRemoveModes(?User $actor, User $user, Group $group, array $modes): self
+    {
+        return static::log($actor, static::USER_REMOVE_MODES, $group, [
+            'details' => [
+                'modes' => $modes,
+                'user_name' => $user->username,
+            ],
+            'user_id' => $user->getKey(),
+        ]);
+    }
+
+
     public static function logUserSetDefault(?User $actor, User $user, Group $group): self
     {
         return static::log($actor, static::USER_SET_DEFAULT, $group, [
+            'details' => [
+                'user_name' => $user->username,
+            ],
             'hidden' => true,
             'user_id' => $user->getKey(),
         ]);
     }
 
     private static function log(?User $actor, string $type, Group $group, array $attributes = []): self
     {
+        $details = [
+            'actor_name' => optional($actor)->username,
+            'group_name' => $group->group_name,
```

Comment:
since it's a historical logging thing I wanted to store the names of the users/groups at this point in time. in the view for group history I'm planning to add later, I'll show `group_name` and `user_name` instead of their current names

I don't know if/how `actor_name` would be used but thought more is better than less, just in case ¯\\\_(ツ)\_/¯

Reply:
Maybe move the assignment for `user_name` here then. Seems duplicated everywhere.